File: jetbrains-plugin/src/main/kotlin/com/asakii/plugin/converters/DisplayItemConverter.kt
package com.asakii.plugin.converters

import com.asakii.plugin.types.*
import com.asakii.claude.agent.sdk.types.*
import kotlinx.serialization.json.*

/**
 * DisplayItem 转换器
 * 
 * 对应 frontend/src/utils/displayItemConverter.ts
 * 将后端的 Message 转换为前端的 DisplayItem
 */
object DisplayItemConverter {
    
    /**
     * 从 ToolUseBlock 创建 ToolCall
     * 
     * @param block 工具使用块
     * @param pendingToolCalls 待处理的工具调用 Map（用于查找已存在的工具调用）
     * @returns ToolCallItem 对象
     */
    fun createToolCall(
        block: ToolUseBlock,
        pendingToolCalls: MutableMap<String, ToolCallItem>
    ): ToolCallItem {
        // 检查是否已存在（用于更新状态）
        val existing = pendingToolCalls[block.id]
        if (existing != null) {
            // 同步更新已存在对象的 input
            // 因为 stream event 中 input_json_delta 会逐步更新 block.input
            return existing
        }
        
        val toolType = ToolConstants.TOOL_NAME_TO_TYPE[block.name] ?: block.name
        val timestamp = System.currentTimeMillis()
        
        // 将 JsonElement 转换为 Map
        val inputMap = jsonElementToMap(block.input)
        
        // 根据工具类型创建具体的 ToolCall
        val toolCall: ToolCallItem = when (toolType) {
            ToolConstants.READ -> ReadToolCall(
                id = block.id,
                timestamp = timestamp,
                status = ToolCallStatus.RUNNING,
                startTime = timestamp,
                input = inputMap
            )
            ToolConstants.WRITE -> WriteToolCall(
                id = block.id,
                timestamp = timestamp,
                status = ToolCallStatus.RUNNING,
                startTime = timestamp,
                input = inputMap
            )
            ToolConstants.EDIT -> EditToolCall(
                id = block.id,
                timestamp = timestamp,
                status = ToolCallStatus.RUNNING,
                startTime = timestamp,
                input = inputMap
            )
            ToolConstants.MULTI_EDIT -> MultiEditToolCall(
                id = block.id,
                timestamp = timestamp,
                status = ToolCallStatus.RUNNING,
                startTime = timestamp,
                input = inputMap
            )
            ToolConstants.TODO_WRITE -> TodoWriteToolCall(
                id = block.id,
                timestamp = timestamp,
                status = ToolCallStatus.RUNNING,
                startTime = timestamp,
                input = inputMap
            )
            ToolConstants.BASH -> BashToolCall(
                id = block.id,
                timestamp = timestamp,
                status = ToolCallStatus.RUNNING,
                startTime = timestamp,
                input = inputMap
            )
            ToolConstants.GREP -> GrepToolCall(
                id = block.id,
                timestamp = timestamp,
                status = ToolCallStatus.RUNNING,
                startTime = timestamp,
                input = inputMap
            )
            ToolConstants.GLOB -> GlobToolCall(
                id = block.id,
                timestamp = timestamp,
                status = ToolCallStatus.RUNNING,
                startTime = timestamp,
                input = inputMap
            )
            ToolConstants.WEB_SEARCH -> WebSearchToolCall(
                id = block.id,
                timestamp = timestamp,
                status = ToolCallStatus.RUNNING,
                startTime = timestamp,
                input = inputMap
            )
            ToolConstants.WEB_FETCH -> WebFetchToolCall(
                id = block.id,
                timestamp = timestamp,
                status = ToolCallStatus.RUNNING,
                startTime = timestamp,
                input = inputMap
            )
            ToolConstants.TASK -> TaskToolCall(
                id = block.id,
                timestamp = timestamp,
                status = ToolCallStatus.RUNNING,
                startTime = timestamp,
                input = inputMap
            )
            ToolConstants.NOTEBOOK_EDIT -> NotebookEditToolCall(
                id = block.id,
                timestamp = timestamp,
                status = ToolCallStatus.RUNNING,
                startTime = timestamp,
                input = inputMap
            )
            ToolConstants.BASH_OUTPUT -> BashOutputToolCall(
                id = block.id,
                timestamp = timestamp,
                status = ToolCallStatus.RUNNING,
                startTime = timestamp,
                input = inputMap
            )
            ToolConstants.KILL_SHELL -> KillShellToolCall(
                id = block.id,
                timestamp = timestamp,
                status = ToolCallStatus.RUNNING,
                startTime = timestamp,
                input = inputMap
            )
            ToolConstants.EXIT_PLAN_MODE -> ExitPlanModeToolCall(
                id = block.id,
                timestamp = timestamp,
                status = ToolCallStatus.RUNNING,
                startTime = timestamp,
                input = inputMap
            )
            ToolConstants.ASK_USER_QUESTION -> AskUserQuestionToolCall(
                id = block.id,
                timestamp = timestamp,
                status = ToolCallStatus.RUNNING,
                startTime = timestamp,
                input = inputMap
            )
            ToolConstants.SKILL -> SkillToolCall(
                id = block.id,
                timestamp = timestamp,
                status = ToolCallStatus.RUNNING,
                startTime = timestamp,
                input = inputMap
            )
            ToolConstants.SLASH_COMMAND -> SlashCommandToolCall(
                id = block.id,
                timestamp = timestamp,
                status = ToolCallStatus.RUNNING,
                startTime = timestamp,
                input = inputMap
            )
            ToolConstants.LIST_MCP_RESOURCES -> ListMcpResourcesToolCall(
                id = block.id,
                timestamp = timestamp,
                status = ToolCallStatus.RUNNING,
                startTime = timestamp,
                input = inputMap
            )
            ToolConstants.READ_MCP_RESOURCE -> ReadMcpResourceToolCall(
                id = block.id,
                timestamp = timestamp,
                status = ToolCallStatus.RUNNING,
                startTime = timestamp,
                input = inputMap
            )
            else -> GenericToolCall(
                id = block.id,
                timestamp = timestamp,
                toolType = toolType,
                status = ToolCallStatus.RUNNING,
                startTime = timestamp,
                input = inputMap
            )
        }
        
        // 添加到 pendingToolCalls
        pendingToolCalls[block.id] = toolCall
        
        return toolCall
    }
    
    /**
     * 更新工具调用结果
     *
     * @param toolCall 工具调用对象
     * @param resultBlock 工具结果块
     */
    @Suppress("KotlinConstantConditions")
    fun updateToolCallResult(toolCall: ToolCallItem, resultBlock: ToolResultBlock): ToolCallItem {
        // 更新状态
        val newStatus = if (resultBlock.isError == true) ToolCallStatus.FAILED else ToolCallStatus.SUCCESS
        val newEndTime = System.currentTimeMillis()

        // 解析结果（content 类型在运行时可能变化，保留分支处理）
        val content = resultBlock.content
        val newResult: ToolResult = if (resultBlock.isError == true) {
            ToolResult.Error(
                error = when (content) {
                    is String -> content
                    is JsonElement -> content.toString()
                    else -> content?.toString() ?: "Unknown error"
                }
            )
        } else {
            ToolResult.Success(
                output = when (content) {
                    is String -> content
                    is JsonElement -> content.toString()
                    else -> content?.toString() ?: ""
                }
            )
        }
        
        // 创建更新后的对象（Kotlin 数据类是不可变的）
        return when (toolCall) {
            is ReadToolCall -> toolCall.copy(status = newStatus, endTime = newEndTime, result = newResult)
            is WriteToolCall -> toolCall.copy(status = newStatus, endTime = newEndTime, result = newResult)
            is EditToolCall -> toolCall.copy(status = newStatus, endTime = newEndTime, result = newResult)
            is MultiEditToolCall -> toolCall.copy(status = newStatus, endTime = newEndTime, result = newResult)
            is TodoWriteToolCall -> toolCall.copy(status = newStatus, endTime = newEndTime, result = newResult)
            is BashToolCall -> toolCall.copy(status = newStatus, endTime = newEndTime, result = newResult)
            is GrepToolCall -> toolCall.copy(status = newStatus, endTime = newEndTime, result = newResult)
            is GlobToolCall -> toolCall.copy(status = newStatus, endTime = newEndTime, result = newResult)
            is WebSearchToolCall -> toolCall.copy(status = newStatus, endTime = newEndTime, result = newResult)
            is WebFetchToolCall -> toolCall.copy(status = newStatus, endTime = newEndTime, result = newResult)
            is TaskToolCall -> toolCall.copy(status = newStatus, endTime = newEndTime, result = newResult)
            is NotebookEditToolCall -> toolCall.copy(status = newStatus, endTime = newEndTime, result = newResult)
            is BashOutputToolCall -> toolCall.copy(status = newStatus, endTime = newEndTime, result = newResult)
            is KillShellToolCall -> toolCall.copy(status = newStatus, endTime = newEndTime, result = newResult)
            is ExitPlanModeToolCall -> toolCall.copy(status = newStatus, endTime = newEndTime, result = newResult)
            is AskUserQuestionToolCall -> toolCall.copy(status = newStatus, endTime = newEndTime, result = newResult)
            is SkillToolCall -> toolCall.copy(status = newStatus, endTime = newEndTime, result = newResult)
            is SlashCommandToolCall -> toolCall.copy(status = newStatus, endTime = newEndTime, result = newResult)
            is ListMcpResourcesToolCall -> toolCall.copy(status = newStatus, endTime = newEndTime, result = newResult)
            is ReadMcpResourceToolCall -> toolCall.copy(status = newStatus, endTime = newEndTime, result = newResult)
            is GenericToolCall -> toolCall.copy(status = newStatus, endTime = newEndTime, result = newResult)
        }
    }
    
    /**
     * 将 Message 数组转换为 DisplayItem 数组（初始化用）
     *
     * @param messages 原始消息数组
     * @param pendingToolCalls 待处理的工具调用 Map
     * @returns DisplayItem 数组
     */
    fun convertToDisplayItems(
        messages: List<com.asakii.claude.agent.sdk.types.Message>,
        pendingToolCalls: MutableMap<String, ToolCallItem>
    ): List<DisplayItem> {
        val displayItems = mutableListOf<DisplayItem>()
        
        for ((messageIdx, message) in messages.withIndex()) {
            when (message) {
                is UserMessage -> {
                    // 用户消息 - content 是 JsonElement
                    val content = parseUserMessageContent(message.content)
                    
                    if (content.isNotBlank()) {
                        val userMessageItem = UserMessageItem(
                            id = generateMessageId(message, messageIdx),
                            content = content,
                            timestamp = System.currentTimeMillis()
                        )
                        displayItems.add(userMessageItem)
                    }
                }
                
                is AssistantMessage -> {
                    // AI 助手消息 - 按顺序处理 content 块
                    // 收集所有文本块的索引，用于标记最后一个文本块
                    val textBlockIndices = message.content
                        .mapIndexedNotNull { idx, block -> if (block is TextBlock && block.text.trim().isNotEmpty()) idx else null }
                    val lastTextBlockIndex = textBlockIndices.lastOrNull() ?: -1
                    
                    for (blockIdx in message.content.indices) {
                        val block = message.content[blockIdx]
                        
                        if (block is TextBlock && block.text.trim().isNotEmpty()) {
                            // 文本块 -> AssistantText
                            val isLastTextBlock = blockIdx == lastTextBlockIndex
                            
                            // 构建统计信息（仅最后一个文本块有）
                            var stats: RequestStats? = null
                            if (isLastTextBlock) {
                                val usage = message.tokenUsage
                                if (usage != null) {
                                    // 查找最近的用户消息时间戳
                                    var lastUserTimestamp = 0L
                                    for (i in messageIdx - 1 downTo 0) {
                                        if (messages[i] is UserMessage) {
                                            lastUserTimestamp = System.currentTimeMillis() // 简化处理
                                            break
                                        }
                                    }
                                    val requestDuration = if (lastUserTimestamp > 0) {
                                        System.currentTimeMillis() - lastUserTimestamp
                                    } else {
                                        0L
                                    }
                                    
                                    stats = RequestStats(
                                        requestDuration = requestDuration,
                                        inputTokens = usage.inputTokens,
                                        outputTokens = usage.outputTokens
                                    )
                                }
                            }
                            
                            val assistantText = AssistantTextItem(
                                id = "${generateMessageId(message, messageIdx)}-text-${displayItems.size}",
                                content = block.text,
                                timestamp = System.currentTimeMillis(),
                                isLastInMessage = isLastTextBlock,
                                stats = stats
                            )
                            displayItems.add(assistantText)
                        } else if (block is ToolUseBlock) {
                            // 工具调用块 -> ToolCall
                            val toolCall = createToolCall(block, pendingToolCalls)
                            displayItems.add(toolCall)
                        }
                    }
                }
                
                is SystemMessage -> {
                    // 系统消息
                    val textContent = message.data.toString()
                    if (textContent.isNotBlank()) {
                        val systemMessageItem = SystemMessageItem(
                            id = generateMessageId(message, messageIdx),
                            content = textContent,
                            level = when (message.subtype) {
                                "error" -> SystemMessageLevel.ERROR
                                "warning" -> SystemMessageLevel.WARNING
                                else -> SystemMessageLevel.INFO
                            },
                            timestamp = System.currentTimeMillis()
                        )
                        displayItems.add(systemMessageItem)
                    }
                }
                
                else -> {
                    // 其他消息类型（忽略或记录日志）
                }
            }
        }
        
        return displayItems
    }
    
    /**
     * 将单个 Message 转换为 DisplayItem 数组（增量更新用）
     *
     * @param message 单个消息
     * @param pendingToolCalls 待处理的工具调用 Map
     * @returns DisplayItem 数组
     */
    fun convertMessageToDisplayItems(
        message: com.asakii.claude.agent.sdk.types.Message,
        pendingToolCalls: MutableMap<String, ToolCallItem>
    ): List<DisplayItem> {
        val displayItems = mutableListOf<DisplayItem>()
        
        when (message) {
            is UserMessage -> {
                // 用户消息 - content 是 JsonElement
                val content = parseUserMessageContent(message.content)
                
                if (content.isNotBlank()) {
                    val userMessageItem = UserMessageItem(
                        id = generateMessageId(message, 0),
                        content = content,
                        timestamp = System.currentTimeMillis()
                    )
                    displayItems.add(userMessageItem)
                }
            }
            
            is AssistantMessage -> {
                // AI 助手消息
                val textBlockIndices = message.content
                    .mapIndexedNotNull { idx, block -> if (block is TextBlock && block.text.trim().isNotEmpty()) idx else null }
                val lastTextBlockIndex = textBlockIndices.lastOrNull() ?: -1
                
                for (blockIdx in message.content.indices) {
                    val block = message.content[blockIdx]
                    
                    if (block is TextBlock && block.text.trim().isNotEmpty()) {
                        val isLastTextBlock = blockIdx == lastTextBlockIndex
                        var stats: RequestStats? = null
                        if (isLastTextBlock) {
                            val usage = message.tokenUsage
                            if (usage != null) {
                                stats = RequestStats(
                                    requestDuration = 0,
                                    inputTokens = usage.inputTokens,
                                    outputTokens = usage.outputTokens
                                )
                            }
                        }
                        
                        val assistantText = AssistantTextItem(
                            id = "${generateMessageId(message, 0)}-text-$blockIdx",
                            content = block.text,
                            timestamp = System.currentTimeMillis(),
                            isLastInMessage = isLastTextBlock,
                            stats = stats
                        )
                        displayItems.add(assistantText)
                    } else if (block is ToolUseBlock) {
                        val toolCall = createToolCall(block, pendingToolCalls)
                        displayItems.add(toolCall)
                    }
                }
            }
            
            else -> {
                // 其他消息类型
            }
        }
        
        return displayItems
    }
    
    /**
     * 生成消息 ID
     */
    private fun generateMessageId(message: com.asakii.claude.agent.sdk.types.Message, index: Int): String {
        return when (message) {
            is UserMessage -> message.sessionId
            is AssistantMessage -> "assistant-$index-${System.currentTimeMillis()}"
            is SystemMessage -> "system-${message.subtype}-$index"
            else -> "message-$index-${System.currentTimeMillis()}"
        }
    }
    
    /**
     * 将 JsonElement 转换为 Map<String, Any?>
     */
    private fun jsonElementToMap(element: JsonElement): Map<String, Any?> {
        if (element !is kotlinx.serialization.json.JsonObject) {
            return emptyMap()
        }
        
        return element.entries.associate { (key, value) ->
            key to jsonElementToAny(value)
        }
    }
    
    /**
     * 将 JsonElement 转换为 Any?
     */
    private fun jsonElementToAny(element: JsonElement): Any? {
        return when {
            element is kotlinx.serialization.json.JsonPrimitive -> {
                when {
                    element.isString -> element.content
                    element.content == "true" -> true
                    element.content == "false" -> false
                    element.content.toIntOrNull() != null -> element.content.toInt()
                    element.content.toLongOrNull() != null -> element.content.toLong()
                    element.content.toDoubleOrNull() != null -> element.content.toDouble()
                    else -> element.content
                }
            }
            element is kotlinx.serialization.json.JsonArray -> {
                element.map { jsonElementToAny(it) }
            }
            element is kotlinx.serialization.json.JsonObject -> {
                jsonElementToMap(element)
            }
            else -> null
        }
    }
    
    /**
     * 解析 UserMessage.content (JsonElement) 为纯文本
     */
    private fun parseUserMessageContent(content: JsonElement): String {
        return when (content) {
            is JsonPrimitive -> {
                // 简单字符串
                content.content
            }
            is JsonArray -> {
                // ContentBlock 数组
                content.mapNotNull { element ->
                    if (element is JsonObject) {
                        val type = element["type"]?.jsonPrimitive?.content
                        when (type) {
                            "text" -> element["text"]?.jsonPrimitive?.content
                            else -> null
                        }
                    } else {
                        null
                    }
                }.joinToString("\n")
            }
            else -> ""
        }
    }
}

File: vscode-extension/src/ide/converters/displayItemConverter.ts
/**
 * DisplayItem 转换器 - 服务端版本
 *
 * 将后端消息格式转换为前端 DisplayItem 格式
 * 主要用于 RSocket 或 HTTP 响应的数据预处理
 *
 * 对应 Kotlin 版本: jetbrains-plugin/src/main/kotlin/com/asakii/plugin/converters/DisplayItemConverter.kt
 */

import type {
    ToolCallItem,
    DisplayItem,
    ToolCallStatus,
    ToolResult,
    UserMessageItem,
    AssistantTextItem,
    SystemMessageItem,
    SystemMessageLevel,
    RequestStats,
} from './types';

/**
 * 工具类型常量
 */
export const TOOL_CONSTANTS = {
    READ: 'Read',
    WRITE: 'Write',
    EDIT: 'Edit',
    MULTI_EDIT: 'MultiEdit',
    TODO_WRITE: 'TodoWrite',
    BASH: 'Bash',
    GREP: 'Grep',
    GLOB: 'Glob',
    WEB_SEARCH: 'WebSearch',
    WEB_FETCH: 'WebFetch',
    TASK: 'Task',
    NOTEBOOK_EDIT: 'NotebookEdit',
    BASH_OUTPUT: 'TaskOutput',
    KILL_SHELL: 'KillShell',
    EXIT_PLAN_MODE: 'ExitPlanMode',
    ASK_USER_QUESTION: 'AskUserQuestion',
    SKILL: 'Skill',
    SLASH_COMMAND: 'SlashCommand',
    LIST_MCP_RESOURCES: 'ListMcpResources',
    READ_MCP_RESOURCE: 'ReadMcpResource',
} as const;

/**
 * 工具名称到类型的映射
 */
export const TOOL_NAME_TO_TYPE: Record<string, string> = {
    'Read': TOOL_CONSTANTS.READ,
    'Write': TOOL_CONSTANTS.WRITE,
    'Edit': TOOL_CONSTANTS.EDIT,
    'MultiEdit': TOOL_CONSTANTS.MULTI_EDIT,
    'TodoWrite': TOOL_CONSTANTS.TODO_WRITE,
    'Bash': TOOL_CONSTANTS.BASH,
    'Grep': TOOL_CONSTANTS.GREP,
    'Glob': TOOL_CONSTANTS.GLOB,
    'WebSearch': TOOL_CONSTANTS.WEB_SEARCH,
    'WebFetch': TOOL_CONSTANTS.WEB_FETCH,
    'Task': TOOL_CONSTANTS.TASK,
    'NotebookEdit': TOOL_CONSTANTS.NOTEBOOK_EDIT,
    'TaskOutput': TOOL_CONSTANTS.BASH_OUTPUT,
    'KillShell': TOOL_CONSTANTS.KILL_SHELL,
    'ExitPlanMode': TOOL_CONSTANTS.EXIT_PLAN_MODE,
    'AskUserQuestion': TOOL_CONSTANTS.ASK_USER_QUESTION,
    'Skill': TOOL_CONSTANTS.SKILL,
    'SlashCommand': TOOL_CONSTANTS.SLASH_COMMAND,
    'ListMcpResources': TOOL_CONSTANTS.LIST_MCP_RESOURCES,
    'ReadMcpResource': TOOL_CONSTANTS.READ_MCP_RESOURCE,
};

/**
 * 工具使用块接口
 */
export interface ToolUseBlock {
    id: string;
    name: string;
    input: Record<string, any>;
}

/**
 * 工具结果块接口
 */
export interface ToolResultBlock {
    tool_use_id: string;
    content?: any;
    isError?: boolean;
    is_error?: boolean;
}

/**
 * 文本块接口
 */
export interface TextBlock {
    type: 'text';
    text: string;
}

/**
 * 内容块联合类型
 */
export type ContentBlock = TextBlock | ToolUseBlock | { type: string; [key: string]: any };

/**
 * Token 使用信息
 */
export interface TokenUsage {
    inputTokens: number;
    outputTokens: number;
}

/**
 * 用户消息接口
 */
export interface UserMessage {
    role: 'user';
    sessionId?: string;
    content: any;
}

/**
 * 助手消息接口
 */
export interface AssistantMessage {
    role: 'assistant';
    content: ContentBlock[];
    tokenUsage?: TokenUsage;
}

/**
 * 系统消息接口
 */
export interface SystemMessage {
    role: 'system';
    subtype?: string;
    data: any;
}

/**
 * 消息联合类型
 */
export type Message = UserMessage | AssistantMessage | SystemMessage;

/**
 * DisplayItem 转换器
 */
export class DisplayItemConverter {
    /**
     * 从 ToolUseBlock 创建 ToolCall
     *
     * @param block 工具使用块
     * @param pendingToolCalls 待处理的工具调用 Map
     * @returns ToolCallItem 对象
     */
    static createToolCall(
        block: ToolUseBlock,
        pendingToolCalls: Map<string, ToolCallItem>
    ): ToolCallItem {
        // 检查是否已存在
        const existing = pendingToolCalls.get(block.id);
        if (existing) {
            return existing;
        }

        const toolType = TOOL_NAME_TO_TYPE[block.name] || block.name;
        const timestamp = Date.now();

        const toolCall: ToolCallItem = {
            id: block.id,
            displayType: 'toolCall',
            toolType,
            toolName: block.name,
            status: 'running' as ToolCallStatus,
            startTime: timestamp,
            timestamp,
            input: block.input || {},
        };

        pendingToolCalls.set(block.id, toolCall);
        return toolCall;
    }

    /**
     * 更新工具调用结果
     *
     * @param toolCall 工具调用对象
     * @param resultBlock 工具结果块
     */
    static updateToolCallResult(
        toolCall: ToolCallItem,
        resultBlock: ToolResultBlock
    ): ToolCallItem {
        const isError = resultBlock.isError === true || resultBlock.is_error === true;
        const newStatus: ToolCallStatus = isError ? 'failed' : 'success';
        const endTime = Date.now();

        const result: ToolResult = isError
            ? { is_error: true, content: resultBlock.content?.toString() || 'Unknown error' }
            : { is_error: false, content: resultBlock.content?.toString() || '' };

        return {
            ...toolCall,
            status: newStatus,
            endTime,
            result,
        };
    }

    /**
     * 将 Message 数组转换为 DisplayItem 数组
     *
     * @param messages 原始消息数组
     * @param pendingToolCalls 待处理的工具调用 Map
     * @returns DisplayItem 数组
     */
    static convertToDisplayItems(
        messages: Message[],
        pendingToolCalls: Map<string, ToolCallItem>
    ): DisplayItem[] {
        const displayItems: DisplayItem[] = [];

        for (let messageIdx = 0; messageIdx < messages.length; messageIdx++) {
            const message = messages[messageIdx];
            const items = this.convertMessageToDisplayItems(
                message,
                pendingToolCalls,
                messageIdx,
                messages
            );
            displayItems.push(...items);
        }

        return displayItems;
    }

    /**
     * 将单个 Message 转换为 DisplayItem 数组
     *
     * @param message 单个消息
     * @param pendingToolCalls 待处理的工具调用 Map
     * @param messageIdx 消息索引
     * @param allMessages 所有消息（用于统计计算）
     * @returns DisplayItem 数组
     */
    static convertMessageToDisplayItems(
        message: Message,
        pendingToolCalls: Map<string, ToolCallItem>,
        messageIdx: number = 0,
        allMessages?: Message[]
    ): DisplayItem[] {
        const displayItems: DisplayItem[] = [];

        switch (message.role) {
            case 'user': {
                const content = this.parseUserMessageContent(message.content);
                if (content.trim()) {
                    const userMessageItem: UserMessageItem = {
                        id: this.generateMessageId(message, messageIdx),
                        displayType: 'userMessage',
                        content,
                        timestamp: Date.now(),
                    };
                    displayItems.push(userMessageItem);
                }
                break;
            }

            case 'assistant': {
                const textBlockIndices = message.content
                    .map((block, idx) => ({ block, idx }))
                    .filter(({ block }) => this.isTextBlock(block) && (block as TextBlock).text.trim())
                    .map(({ idx }) => idx);
                const lastTextBlockIndex = textBlockIndices[textBlockIndices.length - 1] ?? -1;

                for (let blockIdx = 0; blockIdx < message.content.length; blockIdx++) {
                    const block = message.content[blockIdx];

                    if (this.isTextBlock(block) && (block as TextBlock).text.trim()) {
                        const isLastTextBlock = blockIdx === lastTextBlockIndex;

                        let stats: RequestStats | undefined;
                        if (isLastTextBlock && message.tokenUsage) {
                            const usage = message.tokenUsage;
                            stats = {
                                requestDuration: 0,
                                inputTokens: usage.inputTokens,
                                outputTokens: usage.outputTokens,
                            };
                        }

                        const assistantText: AssistantTextItem = {
                            id: `${this.generateMessageId(message, messageIdx)}-text-${blockIdx}`,
                            displayType: 'assistantText',
                            content: (block as TextBlock).text,
                            timestamp: Date.now(),
                            isLastInMessage: isLastTextBlock,
                            stats,
                        };
                        displayItems.push(assistantText);
                    } else if (this.isToolUseBlock(block)) {
                        const toolCall = this.createToolCall(block as ToolUseBlock, pendingToolCalls);
                        displayItems.push(toolCall);
                    }
                }
                break;
            }

            case 'system': {
                const textContent = message.data?.toString() || '';
                if (textContent.trim()) {
                    const level: SystemMessageLevel =
                        message.subtype === 'error' ? 'error' :
                        message.subtype === 'warning' ? 'warning' : 'info';

                    const systemMessageItem: SystemMessageItem = {
                        id: this.generateMessageId(message, messageIdx),
                        displayType: 'systemMessage',
                        content: textContent,
                        level,
                        timestamp: Date.now(),
                    };
                    displayItems.push(systemMessageItem);
                }
                break;
            }
        }

        return displayItems;
    }

    /**
     * 生成消息 ID
     */
    private static generateMessageId(message: Message, index: number): string {
        switch (message.role) {
            case 'user':
                return (message as UserMessage).sessionId || `user-${index}-${Date.now()}`;
            case 'assistant':
                return `assistant-${index}-${Date.now()}`;
            case 'system':
                return `system-${(message as SystemMessage).subtype || 'info'}-${index}`;
            default:
                return `message-${index}-${Date.now()}`;
        }
    }

    /**
     * 解析 UserMessage.content 为纯文本
     */
    private static parseUserMessageContent(content: any): string {
        if (typeof content === 'string') {
            return content;
        }

        if (Array.isArray(content)) {
            return content
                .filter(item => item && typeof item === 'object' && item.type === 'text')
                .map(item => item.text || '')
                .join('\n');
        }

        return '';
    }

    /**
     * 判断是否为文本块
     */
    private static isTextBlock(block: ContentBlock): block is TextBlock {
        return block && (block as any).type === 'text' && typeof (block as TextBlock).text === 'string';
    }

    /**
     * 判断是否为工具使用块
     */
    private static isToolUseBlock(block: ContentBlock): block is ToolUseBlock {
        return block && (block as any).type === 'tool_use' && typeof (block as ToolUseBlock).id === 'string';
    }
}

/**
 * 导出便捷函数
 */
export const createToolCall = DisplayItemConverter.createToolCall.bind(DisplayItemConverter);
export const updateToolCallResult = DisplayItemConverter.updateToolCallResult.bind(DisplayItemConverter);
export const convertToDisplayItems = DisplayItemConverter.convertToDisplayItems.bind(DisplayItemConverter);
export const convertMessageToDisplayItems = DisplayItemConverter.convertMessageToDisplayItems.bind(DisplayItemConverter);
